src/storage/src/render/persist_sink.rs

```diff
@@ -932,439 +935,574 @@ where
+async fn update_persist_with_batches(
+    collection_id: GlobalId,
+    worker_id: usize,
+    shard_id: ShardId,
+    metrics: &SourcePersistSinkMetrics,
+    busy_signal: &Arc<Semaphore>,
+    read_only_rx: &mut watch::Receiver<bool>,
+    bail_on_concurrent_modification: bool,
+    write: &mut mz_persist_client::write::WriteHandle<
+        SourceData,
+        (),
+        mz_repr::Timestamp,
+        StorageDiff,
+    >,
+    upper_cap_set: &mut CapabilitySet<mz_repr::Timestamp>,
+    current_upper: &Rc<std::cell::RefCell<Antichain<mz_repr::Timestamp>>>,
+    source_statistics: &StorageStatistics<
+        SourceStatisticsRecord,
+        SourceStatisticsMetrics,
+        SourceStatisticsMetadata,
+    >,
+    done_batches: &mut Vec<(Antichain<mz_repr::Timestamp>, Antichain<mz_repr::Timestamp>)>,
+    in_flight_batches: &mut HashMap<
+        (Antichain<mz_repr::Timestamp>, Antichain<mz_repr::Timestamp>),
+        BatchSet,
+    >,
+    in_flight_descriptions: &mut std::collections::HashSet<(
+        Antichain<mz_repr::Timestamp>,
+        Antichain<mz_repr::Timestamp>,
+    )>,
+) -> Result<(), anyhow::Error> {
+    let (mut batch_lower, batch_upper) = match (done_batches.first(), done_batches.last()) {
+        (Some(first), Some(last)) => (first.0.clone(), last.1.clone()),
+        (None, None) => {
+            upper_cap_set.downgrade(current_upper.borrow().iter());
+            //
+            return Ok(());
+        }
+        // SAFETY(ptravers): if first exists so must last.
+        (Some(_), None) | (None, Some(_)) => {
+            unreachable!("the list of batches must have a first and last if it is non empty.")
+        }
+    };
+
+    let mut batches: Vec<FinishedBatch> = vec![];
+    let mut batch_metrics: BatchMetrics = BatchMetrics::default();
+    let num_batches = u64::cast_from(done_batches.len());
+
+    for batch_metadata in done_batches.drain(..) {
+        in_flight_descriptions.remove(&batch_metadata);
+        let mut batch = in_flight_batches
+            .remove(&batch_metadata)
+            .unwrap_or_default();
+        batches.append(&mut batch.finished);
+        batch_metrics.add_assign(&batch.batch_metrics);
+    }
 
-                let batch_set = in_flight_batches
-                    .remove(&done_batch_metadata)
-                    .unwrap_or_default();
+    let mut to_append = batches.iter_mut().map(|b| &mut b.batch).collect::<Vec<_>>();
+
+    // NB(ptravers): We loop here to handle competing writers to the same shard.
+    // We handle failure to write in three situations:
+    //  1. the lower of the batch is greater than the current upper of shard - persist_sink exits with an error.
+    //  2. the upper of the shard is less than the upper of the batch and greater
+    //     than the lower of the batch - we retry with a truncated batch.
+    //  3. all other cases - we exit the retry loop after updating the current_upper to the upper reported in the error.
```

Comment:
Is this a failure condition? This is listed as the third item in the "We handle failure to write in three situations" list, but this sounds like the success case to me

Reply:
we failed to append but the shard was moved to the correct upper? so is that a failure?